# Описание проекта

Компания владеет крупным сервисом для чтения книг по подписке. Доступна база данных, содержащая сведения о книгах, издательствах, авторах, а также пользовательских обзораз книг.

**Цель исследования:** проанализировать базу данных для формулирования ценностного предложения для нового продукта. 

### Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

## Получение доступа к БД, обзор таблиц

In [1]:
import pandas as pd
from sqlalchemy import create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'], 
                                                         db_config['pwd'], 
                                                         db_config['host'], 
                                                         db_config['port'],
                                                         db_config['db']) 

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [3]:
def tables_info(table:str):
    '''
    Функция выводит первые 5 строк таблиц table из базы данных,
    а также сводную информацию о таблице.
    '''
    query = f' \
    SELECT * \
    FROM {table}'
    
    df = pd.io.sql.read_sql(query, con = engine)
    print(f'ТАБЛИЦА {table}')
    display(df.head())
    df.info()
    print('------------------')

In [4]:
tables = ['books', 'authors', 'publishers', 'ratings', 'reviews']
for i in tables:
    tables_info(i)

ТАБЛИЦА books


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB
------------------
ТАБЛИЦА authors


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB
------------------
ТАБЛИЦА publishers


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB
------------------
ТАБЛИЦА ratings


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB
------------------
ТАБЛИЦА reviews


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB
------------------


Данные выгружаются корректно, пропуски отсутствуют.

In [5]:
# функция для вывода query запросов к БД
def get_data(query:str):
    df = pd.io.sql.read_sql(query, con = engine)
    return df

## Решение задач

### Посчитайте, сколько книг вышло после 1 января 2000 года

In [8]:
get_data('''

SELECT COUNT(*)
FROM books
WHERE publication_date > '2000-01-01'

''')

count
0    819

После 1 января 2000 года вышло 819 книг.

### Для каждой книги посчитайте количество обзоров и среднюю оценку

In [10]:
get_data('''

WITH 
-- Таблица со средней оценкой
avg_rating AS
  (SELECT book_id,
          AVG(rating) AS average_rating
   FROM ratings
   GROUP BY book_id), 
   
-- Таблица с кол-вом обзоров
n_of_reviews AS
  (SELECT book_id,
          COUNT (*) AS number_of_reviews
   FROM reviews
   GROUP BY book_id) 
   
-- Общая таблица с названиями книг
SELECT b.book_id,
       b.title,
       n_of_reviews.number_of_reviews,
       avg_rating.average_rating
FROM books AS b
LEFT JOIN avg_rating ON b.book_id = avg_rating.book_id
LEFT JOIN n_of_reviews ON b.book_id = n_of_reviews.book_id

''')



book_id                                              title  \
0        652          The Body in the Library (Miss Marple  #3)   
1        273                                          Galápagos   
2         51                           A Tree Grows in Brooklyn   
3        951  Undaunted Courage: The Pioneering First Missio...   
4        839                                        The Prophet   
..       ...                                                ...   
995      672   The Cat in the Hat and Other Dr. Seuss Favorites   
996       83    Anne Rice's The Vampire Lestat: A Graphic Novel   
997      221                          Essential Tales and Poems   
998      387                               Leonardo's Notebooks   
999      808                            The Natural Way to Draw   

     number_of_reviews  average_rating  
0                  2.0        4.500000  
1                  2.0        4.500000  
2                  5.0        4.250000  
3                  2.0        4.000000  
4                  4.0        4.285714  
..                 ...             ...  
995                NaN        5.000000  
996                NaN        3.666667  
997                NaN        4.000000  
998                NaN        4.000000  
999                NaN        3.000000  

[1000 rows x 4 columns]

Количество обзоров и средние оценки варьируются. По ряду книг обзоры не написаны, несмотря на хорошие пользовательские оценки.

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц

In [11]:
get_data('''

SELECT p.publisher,
       COUNT(*) AS number_of_books
FROM books AS b
LEFT JOIN publishers AS p ON b.publisher_id=p.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher
ORDER BY number_of_books DESC
LIMIT 1
''')

publisher  number_of_books
0  Penguin Books               42

Издательство Penguin Books выпустило наибольшее количество книг толще 50 стр. (42 шт.).

### Определите автора с самой высокой средней оценкой книг (учитываются только книги с 50 и более оценками)

In [13]:
get_data('''

WITH 
-- Книги с 50 и более оценками
top_books AS
  (SELECT book_id,
          AVG(rating) AS average_rating
   FROM ratings
   GROUP BY book_id
   HAVING COUNT(rating_id) >= 50) 
   
-- Добавляем авторов, группируем по ним, считаем среднюю оценку
SELECT a.author,
       AVG(top_books.average_rating) AS average_rating
FROM books AS b
JOIN top_books ON top_books.book_id = b.book_id
JOIN authors AS a ON b.author_id = a.author_id
GROUP BY author
ORDER BY average_rating DESC
LIMIT 1

''')

author  average_rating
0  J.K. Rowling/Mary GrandPré        4.283844

J.K. Rowling (совместно с Mary GrandPré) - автор с наиболее высокой средней оценкой книг (веротяно, книги о Гарри Поттере).

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [14]:
get_data('''

WITH users AS
  (SELECT username,
          COUNT(rating_id) AS number_of_ratings
   FROM ratings
   GROUP BY username
   HAVING COUNT(rating_id) > 50)
   
SELECT AVG(i.number_of_reviews)
FROM
  (SELECT COUNT(r.review_id) AS number_of_reviews
   FROM reviews AS r
   RIGHT JOIN users ON r.username = users.username
   GROUP BY users.username) AS i


''')

avg
0  24.333333

Пользователи, поставившие более 50 оценок, в среднем пишут по 24 обзора.

## Итоговый вывод

В базе содержится информация о 1000 книг, большинство из которых вышло после 1 января 2000 года. По ряду книг обзоры не написаны, несмотря на хорошие пользовательские оценки. Лидер по выпуску книг - издательство Penguin Books. Наиоболее рейтингованный писатель - J.K. Rowling. Пользователи, которые любят ставить оценки, часто пишут обзоры.